src/agents/experimental/streamlit/agent.ts

```diff
@@ -0,0 +1,181 @@
+  public emitter = new Emitter<Events>({
+    namespace: ["agent", "experimental", "streamlit"],
+    creator: this,
+  });
+
+  constructor(protected readonly input: Input) {
+    super();
+  }
+
+  public get meta(): AgentMeta {
+    return {
+      name: `Streamlit`,
+      tools: [],
+      description: `StreamlitAgent is an experimental meta-app agent that uses \`Meta LLaMa 3.1 70B\` to build \`IBM Granite 3 8B\`-powered apps using Streamlit -- a simple UI framework for Python.`,
+    };
+  }
+
+  public get memory() {
+    return this.input.memory;
+  }
+
+  protected async _run(
+    input: RunInput,
+    _options: Options | undefined,
+    run: GetRunContext<typeof this>,
+  ): Promise<RunOutput> {
+    const { userMessage, runMemory } = await this.prepare(input);
+
+    let content = "";
+    for await (const chunk of this.input.llm.stream(runMemory.messages, {
+      signal: run.signal,
+    })) {
+      const delta = chunk.getTextContent();
+      content += delta;
+      await run.emitter.emit("newToken", { delta, state: { content }, chunk });
+    }
+    const result = this.parse(content);
+
+    const assistantMessage = BaseMessage.of({
+      role: Role.ASSISTANT,
+      text: content,
+    });
+    await this.memory.addMany([userMessage, assistantMessage].filter(isTruthy));
+
+    return {
+      result,
+      message: assistantMessage,
+      memory: runMemory,
+    };
+  }
+
+  protected async prepare(input: RunInput) {
+    const systemMessage = BaseMessage.of({
+      role: Role.SYSTEM,
+      text: (this.input.templates?.system ?? StreamlitAgentSystemPrompt).render({}),
+    });
+
+    const userMessage = input.prompt
+      ? BaseMessage.of({
+          role: Role.USER,
+          text: input.prompt,
+          meta: {
+            createdAt: new Date(),
+          },
+        })
+      : null;
+
+    const runMemory = new TokenMemory({
+      llm: this.input.llm,
+      capacityThreshold: 0.85,
+      syncThreshold: 0.6,
+      handlers: {
+        removalSelector(msgs) {
+          const msgToRemove =
+            msgs.length > 3 &&
+            (msgs.find((msg) => msg.role === Role.ASSISTANT) ??
+              msgs.find((msg) => msg.role === Role.USER));
+
+          if (!msgToRemove || (msgToRemove.role === Role.USER && msgs.at(-1) === msgToRemove)) {
+            throw new AgentError("Cannot fit the current conversation into the context window!");
+          }
+          return msgToRemove;
+        },
+      },
+    });
+    await runMemory.addMany(
+      [systemMessage, ...this.input.memory.messages, userMessage].filter(isTruthy),
+    );
+
+    return { runMemory, userMessage };
+  }
+
+  protected parse(content: string): Result {
+    const match = findFirstPair(content, ["```python-app\n", "```\n"]);
+    return {
+      raw: content,
+      app: match ? match.inner : null,
+      text: match
+        ? `${content.substring(0, match.start)}${content.substring(match.end + 1)}`
```

Comment:
This expects only one python-app block to be present, but from my experience the agent sometimes writes multiple code blocks with the last one usually being the one you want to use.

Reply:
Also the information where the code block was is lost in `text`. It would probably work better if we returned a structure like

```
[
    { type: "text", content: "..." },
    { type: "app", content: "..." },
    { type: "text", content: "..." },
    // ...
]
```

Then the consumer could use the code blocks as they see fit.

Reply:
Updated.